pkg/sql/schemachanger/scplan/internal/scstage/build.go

```diff
@@ -563,6 +628,26 @@ func (bc buildContext) nodes(current []scpb.Status) []*screl.Node {
 	return nodes
 }
 
+// computeTargetCanStillFailMap pre-computes for each node whether any op
+// edges reachable from the node are non-mutation ops (backfills or
+// validations) and thus can fail; mutation operations cannot fail.
+func computeTargetCanStillFailMap(g *scgraph.Graph, init scpb.CurrentState) map[*screl.Node]bool {
+	m := make(map[*screl.Node]bool, g.Order())
+	var rec func(n *screl.Node) bool
+	rec = func(n *screl.Node) (ret bool) {
+		defer func() { m[n] = ret }()
+		oe, ok := g.GetOpEdgeFrom(n)
+		if !ok {
+			return false
+		}
+		return rec(oe.To()) || oe.Type() != scop.MutationType
+	}
+	for _, n := range nodes(g, init.Targets, init.Current) {
+		rec(n)
+	}
+	return m
+}
```

Comment:
Ah. This makes sense. 

This logic belongs in the bowels of opgen, though, along with what determines the `revertible` boolean field on each op-edge. You'll find that it's quite similar. 

In fact, perhaps this `revertible` boolean field should be replaced by an enum with three states:
- revertible, failure possible;
- revertible, failure not possible;
- not revertible, failure not possible.

Reply:
I moved this into opgen in the newly added first commit. It was the right thing to do, but took a little bit of elbow grease. 

I've left two booleans as I found it easier to deal with for now, do you feel strongly?

Reply:
No, I don't feel strongly about that.